include/zephyr/rtio/rtio.h

```diff
@@ -0,0 +1,418 @@
+ * The kernel is the consumer of the submission queue, and producer of the completion queue.
+ * The application is the consumer of the completion queue and producer of the submission queue.
+ *
+ * Nothing is done until a call is performed to do the work (rtio_execute).
+ */
+struct rtio {
+
+	/*
+	 * An executor which does the job of working through the submission
+	 * queue.
+	 */
+	struct rtio_executor *executor;
+
+	/* Number of completions that were unable to be submitted with results
+	 * due to the cq spsc being full
+	 */
+	atomic_t xcqcnt;
+
+	/* Submission queue */
+	struct rtio_sq *sq;
+
+	/* Completion queue */
+	struct rtio_cq *cq;
+};
+
+/**
+ * @brief RTIO Submission Flags
+ * @defgroup rtio_submission_flags RTIO Submission Flags
+ * @ingroup rtio_api
+ * @{
+ */
+
+/**
+ * @brief Submit a request to be done while blocking the caller
+ *
+ * The submitted operation will be done synchronously and the call shall
+ * not return until the operation is complete. The default is a non-blocking
+ * request.
+ */
+#define RTIO_SUBMIT_SYNC BIT(0)
+
+/**
+ * @}
+ */
+
+
+/**
+ * @brief API that an RTIO IO device should implement
+ */
+struct rtio_iodev_api {
+	/**
+	 * @brief Submission function for a request to the iodev
+	 *
+	 * By default with no flags provided the function should
+	 * return as soon as possible. It is acceptable but not
+	 * ideal if the function must poll on hardware for completion
+	 * of the request.
+	 *
+	 * If the RTIO_SUBMIT_SYNC flag is set the function MUST
+	 * complete the request before returning.
+	 *
+	 * Since the entire context (including the executor) is
+	 * passed in the executor may be used in conjunction with the
+	 * requested operation in a implementation defined way. For
+	 * example to setup a DMA transfer.
+	 */
+	void (*submit)(const struct rtio_sqe *sqe,
+			struct rtio *r,
+			uint32_t flags);
+};
+
+/**
+ * @brief An IO device with a function table for submitting requests
+ *
+ * This is required to be the first member of every iodev. There's a strong
+ * possibility this will be extended with some common data fields (statistics)
+ * in the future.
+ */
+struct rtio_iodev {
+	const struct rtio_iodev_api * const api;
+};
+
+
+/** An operation that does nothing and will complete immediately */
+#define RTIO_OP_NOP 0
+
+/** An operation that receives (reads) */
+#define RTIO_OP_RX 1
+
+/** An operation that transmits (writes) */
+#define RTIO_OP_TX 2
+
+/**
+ * @brief Define a fixed length submission queue
+ *
+ * @param name Variable name
+ * @param len Queue length, power of 2 required (2, 4, 8)
+ */
+#define RTIO_SQ_DEFINE(name, len)			\
+	RTIO_SPSC_DEFINE(name, struct rtio_sqe, len)
+
+/**
+ * @brief Acquire and prepare a nop (no op) request
+ */
+static inline void rtio_sqe_prep_nop(struct rtio_sqe *sqe,
+				     struct rtio_iodev *iodev, void *userdata)
+{
+	sqe->op = RTIO_OP_NOP;
+	sqe->iodev = iodev;
+	sqe->userdata = userdata;
+}
+
+/**
+ * @brief Acquire and prepare a read request
```

Comment:
where is the aquire portion of this?

Reply:
Originally this acquired from the sq and prepped, I've since corrected this and the other similar doc strings in an updated commit